### PR TITLE
feat: remove model key reconciliation from auth worker

### PR DESCRIPTION
**Pull request by SimoneDutto**
# Description

Remove the authorized keys model reconciliation, now everything is proxyed through the controller, so user's keys should not be propagated to the user.  Now the authentication worker just manages ephemeral keys.
Contextually i've also removed unused authorized key client, and the associated bash suite.

# QA
`make install`

bootstrap the controller

For QA i need to give some additional context: during bootstrap you pub keys are added to the model in juju's state and to the actual machine manually (this will be addressed in another pr soon).
Before this patch, even if you removed your host key from the machine they will be back in the model because there was the reconciliation logic putting model keys back into the machine.

Now with this patch if you remove these keys, they will not come back.

```
juju switch controller
juju ssh 0
cat .ssh/authorized_keys 
# you should see the ephemeral key + your host keys
# now edit the keys to remove your host keys
# they should not reappear
```

### CODE REVIEW

File: api/connection.go [hunk 1]
```diff
@@ -237,11 +236,6 @@ func addAddress(servers []network.MachineHostPorts, addr string) ([]network.Mach
 	return result, nil
 }
 
-// KeyUpdater returns access to the KeyUpdater API
-func (c *conn) KeyUpdater() *keyupdater.Client {
-	return keyupdater.NewClient(c)
-}
```
Comment (kian99) [inline api/connection.go:243]: How come you could remove this?

**Review by kian99 — COMMENTED** [review]
I expect there would also be some domain code you can remove.

File: api/connection.go (hunk 1 above)
Comment (SimoneDutto) [reply api/connection.go:243]: it had no uses other that in that worker

**Comment by SimoneDutto** [comment]
> I expect there would also be some domain code you can remove.

not really because we need to keep the facade around for older agents during upgrades. But i'd need @SimonRichardson to confirm if that's the case

**Comment by SimoneDutto** [comment]
> ## Code Review Summary
> **Verdict: Comment** (1 issue, 2 suggestions)
> 
> ### Critical
> 1. `cmd/jujuagentd/agent/machine/manifolds_test.go:2144` — The `expectedMachineManifoldsWithDependenciesIAAS` map still lists `"ssh-authkeys-updater"` with dependencies `"agent"`, `"api-caller"`, `"api-config-watcher"`, `"migration-fortress"`, `"migration-inactive-flag"`, `"upgrade-check-flag"`, `"upgrade-check-gate"`, `"upgrade-steps-flag"`, `"upgrade-steps-gate"`. After this PR, the authentication worker manifold has **zero inputs** (confirmed by `manifold_test.go` asserting `tc.HasLen, 0`). The `TestManifoldsDependenciesIAAS` test will fail because `ManifoldDependencies` traverses the `Inputs` field, which is now empty. The `ssh-authkeys-updater` entry in this map must be changed to an empty slice: `"ssh-authkeys-updater": {}`. Note that `ssh-session` (line 2161) still correctly lists `"ssh-authkeys-updater"` as a dependency — that relationship is unchanged.

Addressed

> ### Warnings
> 1. `internal/worker/authenticationworker/worker_test.go` — The `TestRemoveLastEphemeralKey` test was removed. This covered an important edge case: removing the only key from `authorized_keys` when the file has no other keys. The underlying code handling this (`ssh.DeleteKeysFromFile` in `handleEphemeralRequest`, used instead of `ssh.DeleteKeys` which refuses to delete the last key) is still present and unchanged. This edge case should still be tested. Re-add a test that starts a worker with no model keys, adds a single ephemeral key, and verifies that removing it succeeds and leaves an empty `authorized_keys` file.
> 
> ### Suggestions
> 1. `apiserver/facades/agent/keyupdater/authorisedkeys.go` — The server-side `KeyUpdater` facade (with `AuthorisedKeys` and `WatchAuthorisedKeys` methods) and its registration in `api/facadeversions.go` are not removed by this PR. Since the client-side `api/agent/keyupdater` package is fully deleted, no current juju client calls this facade. If the facade is retained for backward compatibility with older agents during upgrades, add a comment explaining this. Otherwise, the facade and its service registration should be removed in a follow-up PR to avoid dead server-side code.

Added

> 2. `internal/worker/authenticationworker/worker.go` — The `authorizedKeysFile` constant (line 22) is still defined and used in `handleEphemeralRequest` for `ssh.DeleteKeysFromFile`. This is fine, but consider whether `SSHUser` (line 20, currently `"ubuntu"`) is still meaningful now that the worker no longer reconciles model keys. If the worker only manages ephemeral keys for reverse tunnels, the SSH user is only used for `ssh.AddKeys`/`ssh.DeleteKeysFromFile` calls — verify this is still the correct user for ephemeral key management.

This is correct.

> ### Architecture and Alternative Approaches
> * The simplification of the manifold from `Manifold(config ManifoldConfig, output)` to `Manifold(output)` is clean and appropriate. Removing the `engine.AgentAPIManifoldConfig` indirection and the `Client` interface eliminates unnecessary complexity now that the worker has no API dependencies. The `dependency.Manifold{Start: newWorker}` pattern with no inputs is the correct minimal form for a self-contained worker.
> * The `ManifoldConfig` type alias (`type ManifoldConfig engine.AgentAPIManifoldConfig`) is correctly removed — it was a pure pass-through with no additional fields or validation.
> * Consider renaming the worker type from `AuthWorker` to something more specific like `EphemeralKeyWorker` in a follow-up, since the worker no longer handles authentication broadly but only ephemeral SSH key management.
> 
> ### Looks Good
> * Clean removal of the model key reconciliation logic (`setUp`, `handleModelKeyUpdate`, `writeSSHKeys`) — the remaining `loop()` is minimal and correct, handling only catacomb lifecycle and ephemeral key requests.
> * The `handleEphemeralRequest` function is correctly retained unchanged, including the `DeleteKeysFromFile` usage for the last-key edge case.
> * The `enqueue` method with its `ErrAuthenticationWorkerDying` sentinel is correctly preserved, maintaining the contract that the sshsession worker relies on.
> * Import cleanup in `api/connection.go`, `api/package_test.go`, `cmd/containeragent/initialize/package_test.go`, and `cmd/containeragent/unit/package_test.go` is thorough and correct.
> * Removing the integration test `test_machine_ssh` from `tests/suites/authorized_keys/task.sh` and deleting `machine.sh` is appropriate since model key propagation to machines is no longer a worker responsibility.
> * The updated `worker_test.go` correctly tests the remaining functionality: add/remove ephemeral keys and the dying-error contract.
> 
> Reviewed by lee-van-cleef[bot]

**Comment by kian99** [comment]
> > I expect there would also be some domain code you can remove.
> 
> not really because we need to keep the facade around for older agents during upgrades. But i'd need @SimonRichardson to confirm if that's the case

Since this is going into 4.1, the only way you could get an old agent trying to use the removed facade is during model migration. But I don't believe that would prevent the model from being upgraded and then you get the new agent on all machines/units.

But let's wait for confirmation.

File: apiserver/facades/agent/keyupdater/authorisedkeys.go [hunk 2]
```diff
@@ -20,7 +20,8 @@ import (
 )
 
 // KeyUpdaterAPI implements the KeyUpdater interface and is the concrete
-// implementation of the api end point.
+// implementation of the API endpoint. It remains registered for compatibility
+// with older agents during upgrades; current agents no longer consume it.
```
Comment (SimonRichardson) [inline apiserver/facades/agent/keyupdater/authorisedkeys.go]: This is a shame, because if we did this earlier and got 3.6 and 4.0 to handle upgrades when the facade was returning `not supported`, we could have removed this wholesale.

File: internal/worker/authenticationworker/worker.go [hunk 3]
```diff
@@ -4,21 +4,13 @@
 package authenticationworker
```
Comment (SimonRichardson) [inline internal/worker/authenticationworker/worker.go]: I wonder if we should rename this now.

**Review by SimonRichardson — APPROVED** [review]
I think we should rename the worker before landing, but otherwise this is good.

File: internal/worker/authenticationworker/worker.go (hunk 3 above)
Comment (SimoneDutto) [reply internal/worker/authenticationworker/worker.go]: i think the worker name is still holding good because 
`authenticationWorkerName           = "ssh-authkeys-updater"`

we are still updating auth keys

File: apiserver/facades/agent/keyupdater/authorisedkeys.go (hunk 2 above)
Comment (SimoneDutto) [reply apiserver/facades/agent/keyupdater/authorisedkeys.go]: ah ok, so you confirm this https://github.com/juju/juju/pull/23184#issuecomment-5507346781?

Comment (SimonRichardson) [reply apiserver/facades/agent/keyupdater/authorisedkeys.go]: Yeah, as handleModelKeyUpdate, just exits if it encounters an error, which will cause this to bounce. This might not be such a problem in the grand scheme of things, but it *may* prevent the upgrade from happening. You could run an experiment of running a 3.6 or 4.0 to main migration, which should automatically upgrade, without the facades!

File: internal/worker/authenticationworker/worker.go (hunk 3 above)
Comment (SimonRichardson) [reply internal/worker/authenticationworker/worker.go]: I was thinking it should be `sshauthkeysupdater` rather than `authenticationworker`.

File: apiserver/facades/agent/keyupdater/authorisedkeys.go (hunk 2 above)
Comment (SimoneDutto) [reply apiserver/facades/agent/keyupdater/authorisedkeys.go]: even if the experiement is successful i'd be scared that some conditions might bounce other workers during upgrade. So i'm inclined to leave it in there, it's not a huge deal, worst case is dead code, not even that much

File: internal/worker/authenticationworker/worker.go (hunk 3 above)
Comment (SimoneDutto) [reply internal/worker/authenticationworker/worker.go]: sounds good

Comment (SimoneDutto) [reply internal/worker/authenticationworker/worker.go]: done

File: apiserver/facades/agent/keyupdater/authorisedkeys.go (hunk 2 above)
Comment (SimonRichardson) [reply apiserver/facades/agent/keyupdater/authorisedkeys.go]: Might be worth stating that we could remove it in later additions, once we know upgrades are safe.

Comment (SimoneDutto) [reply apiserver/facades/agent/keyupdater/authorisedkeys.go]: from analyzing the code and the help of our AI friend:
- the authentication worker in juju 3 and juju 4 is wrapped by `ifNotMigrating`. 
- However an old agent could still race with the flag to try and create a key updater client
- if that happens the worker could fail to start and set the migration flag
- leaving the old agent in an error state

So, to be caution and since i don't have a way to produce the race reliably, i'll leave this facade here with a comment explaining that it could be later removed

File: internal/worker/authenticationworker/worker.go (hunk 3 above)
Comment (kian99) [reply internal/worker/authenticationworker/worker.go]: Regarding naming, just leaving a note that we still have the very old `hostkeyreporter` worker which reports machine's host keys to the controller and now a `sshauthkeysupdater` worker which puts ephemeral host keys into machine's authorized keys file.

**Review by kian99 — APPROVED** [review]
We should rename all places that were using the authentication worker like inside `internal/worker/sshsession/manifold.go`, we use the authentication worker,
```go
	// AuthenticationWorkerName is the dependency name of the worker that
	// manages the machine's authorized_keys file, including ephemeral keys.
	AuthenticationWorkerName string
```

**Review by kian99 — COMMENTED** [review]
Nice I see you just did it
